Add support for FreeBSD

tagattie opened the pull request:
<!-- Thank you for submitting a Pull Request. Please:
* Read our Pull Request guidelines:
  https://github.com/microsoft/vscode/wiki/How-to-Contribute#pull-requests
* Associate an issue with the Pull Request.
* Ensure that the code is up-to-date with the `main` branch.
* Include a description of the proposed changes and how to test them.
-->

This PR contains a set of changes for adding support for FreeBSD.

Basically, FreeBSD can be treated as if it is Linux. So the changes are mainly adding `process.platform === 'freebsd'`, `process.platform !== 'freebsd'`, or `case 'freebsd'`: to wherever it checks the OS is Linux with a few exceptions where the check is for really Linux-specific.

I would much appreciate it if this PR is considered for merging.

Relate PR: https://github.com/microsoft/vscode-deviceid/pull/25
Related Issues: #1359, #30046, #32409, #74857, #129928

tagattie committed:
Add support for FreeBSD

Basically FreeBSD can be treated as if it is Linux. So the changes are just adding process.platform === 'freebsd', process.platform !== 'freebsd', or case 'freebsd': to wherever it checks the OS is Linux with few exceptions where the check is for really Linux-specific.

tagattie committed:
Suppress output to stderr

There are no /proc/version file in the FreeBSD's procfs. So redirect "No such file or directory" message to /dev/null.

tagattie committed:
Skip glibc check on FreeBSD

FreeBSD is not a glibc-based system.

tagattie committed:
Adjust arguments to ps command

The format '-o pid=,ppid=,pcpu=,pmem=,command=' is not supported by FreeBSD. So adjust it so that the arguments can be accepted by ps command on both Linux and FreeBSD.

benz0li commented:
@joaomoreno Please consider this PR. It seems to make @tagattie life easier to maintain the FreeBSD port.

He is the official maintainer of https://www.freshports.org/editors/vscode.

FYI @bpasero

tagattie committed:
Add one missing patch to support FreeBSD

First I thought this change is unnecessary because there is not update service provided for FreeBSD. However, without this change code-oss does not start with error:
Error: [invokeFunction] unknown service 'updateService'

joaomoreno commented:
I'm really sorry, but this isn't something we're committed to doing. All the issues linked above are closed as "not planned" or "out of scope" for a reason. FreeBSD support isn't just what this PR does. It's a long term commitment into maintaining support within the codebase,  periodically testing VS Code on that platform, updating and maintaining all our documentation, ideally having someone who selfhosts full time, etc.

Thank you for the changes anyway, they serve as a good base if eventually this becomes a priority for us.

benz0li commented:
@joaomoreno Thank you for the clarification.
